### PR TITLE
Fix race condition when waiting for a concurrent layer pull

### PR DESCRIPTION
**Pull request by aaronlehmann**
Before, this only waited for the download to complete. There was no guarantee that the layer had been registered in the graph and was ready use. This is especially problematic with v2 pulls, which wait for all downloads before extracting layers.

Change Broadcaster to allow an error value to be propagated from Close to the waiters.

Make the wait stop when the extraction is finished, rather than just the download.

This also fixes v2 layer downloads to prefix the pool key with "layer:" instead of "img:". "img:" is the wrong prefix, because this is what v1 uses for entire images. A v1 pull waiting for one of these operations to finish would only wait for that particular layer, not all its dependencies.

cc @docker/distribution-maintainers @tonistiigi

### CODE REVIEW

File: graph/pull_v2.go [hunk 1]
```diff
@@ -74,14 +74,22 @@ func (p *v2Puller) pullV2Repository(tag string) (err error) {
...
+		result := broadcaster.Wait()
+		if err, ok := result.(error); ok {
+			return err
+		}
 		return nil
 	}
-	defer p.poolRemove("pull", taggedName)
-	broadcaster.Add(p.config.OutStream)
+	defer func() {
+		if err != nil {
```
Comment (dmcgowan) [inline graph/pull_v2.go]: This block is equivalent to just doing `p.poolRemove("pull", taggedName, err)`

Comment (dmcgowan) [inline graph/pull_v2.go]: Actually is it, saw it expects `interface{}`, if you pass in a nil error vs nil interface{} is it treated the same? If this is required should add a comment.

Comment (aaronlehmann) [inline graph/pull_v2.go]: I believe that if this block was replaced with `p.poolRemove("pull", taggedName, err)`, then in the case where `err` is nil, the type switch cases that check for an `error` would succeed. In other words, a nil error is different than a nil `interface{}`. I'd love suggestions of a cleaner way to deal with this. I agree it's ugly.

Comment (aaronlehmann) [inline graph/pull_v2.go]: https://golang.org/doc/faq#nil_error

Comment (dmcgowan) [inline graph/pull_v2.go]: The only way I see to clean it up would be to explicitly use the `error` type.  Right now I am trying to understand the necessity of the `Wait` calls which return a broadcaster.

**Comment by aaronlehmann** [comment]
This was overly complicated because it tried to protect against a deadlock that @dmcgowan convinced me can't actually happen. Repushed a simpler version that only uses a single broadcaster, and avoids using the `interface{}` type when passing data to waiters on broadcasters.

**Comment by dmcgowan** [comment]
LGTM

**Comment by aaronlehmann** [comment]
Added two small related fixes to this PR:

- Fix a data race accessing d.poolKey and d.tmpFile in the `defer`ed cleanup code. Preinitialize these values in the same goroutine that runs the cleanup instead of letting it happen concurrently.
- Remove the `io.Pipe` which was added in #15353. It's no longer necessary, because the `Broadcaster` which downloading goroutines write to now serves the same purpose. When `pullV2Tag` returns, all broadcasters are closed, and subsequent writes to them return errors.

cc @jlhawn

**Comment by cpuguy83** [comment]
Needs a rebase.

**Comment by aaronlehmann** [comment]
Rebased.

File: graph/pull_v1.go [hunk 2]
```diff
@@ -225,8 +232,9 @@ func (p *v1Puller) pullRepository(askedTag string) error {
 	return nil
 }
 
-func (p *v1Puller) pullImage(out io.Writer, imgID, endpoint string, token []string) (bool, error) {
-	history, err := p.session.GetRemoteHistory(imgID, endpoint)
+func (p *v1Puller) pullImage(out io.Writer, imgID, endpoint string, token []string) (layersDownloaded bool, err error) {
+	var history []string
```
Comment (cpuguy83) [inline graph/pull_v1.go]: What's the purpose of this change?

Comment (aaronlehmann) [inline graph/pull_v1.go]: The defer on line 267 needs to see if the function is returning an error, so this function was changed to use named return values.

Comment (cpuguy83) [inline graph/pull_v1.go]: Oh, this is so err isn't redefined. Gotcha.

File: pkg/progressreader/broadcaster.go [hunk 3]
```diff
@@ -130,14 +133,16 @@ func (broadcaster *Broadcaster) Add(w io.Writer) error {
 	return nil
 }
 
-// Close signals to all observers that the operation has finished.
-func (broadcaster *Broadcaster) Close() {
+// Close signals to all observers that the operation has finished. Its argument
+// is a result that should be returned to waiters blocking on Wait.
+func (broadcaster *Broadcaster) Close(result error) {
```
Comment (stevvooe) [inline pkg/progressreader/broadcaster.go]: Typically, we call methods with this signature `CloseWithError`.

File: graph/tags.go [hunk 4]
```diff
@@ -462,18 +462,18 @@ func (store *TagStore) poolAdd(kind, key string) (*progressreader.Broadcaster, b
 	return broadcaster, false
 }
 
-func (store *TagStore) poolRemove(kind, key string) error {
+func (store *TagStore) poolRemove(kind, key string, broadcasterResult error) error {
```
Comment (stevvooe) [inline graph/tags.go]: If we are going to have this method called with a `nil` argument everywhere, let's have a version called `poolRemoveWithError`.

File: graph/pull_v1.go [hunk 5]
```diff
@@ -138,16 +138,22 @@ func (p *v1Puller) pullRepository(askedTag string) error {
...
+					errors <- err
+					return
+				}
 				out.Write(p.sf.FormatProgress(stringid.TruncateID(img.ID), "Download complete", nil))
 				errors <- nil
 				return
 			}
-			broadcaster.Add(out)
-			defer p.poolRemove("pull", "img:"+img.ID)
+			defer func() {
```
Comment (dmcgowan) [inline graph/pull_v1.go]: extra closure no longer needed

**Comment by dmcgowan** [comment]
LGTM

**Comment by stevvooe** [comment]
LGTM

**Comment by cpuguy83** [comment]
Needs a rebase.

**Comment by aaronlehmann** [comment]
Rebased.

**Comment by cpuguy83** [comment]
Test failure seems like it may be legit?

File: graph/pull_v1.go [hunk 6]
```diff
@@ -138,16 +138,20 @@ func (p *v1Puller) pullRepository(askedTag string) error {
...
 
 			// ensure no two downloads of the same image happen at the same time
-			broadcaster, found := p.poolAdd("pull", "img:"+img.ID)
+			poolKey := "img:" + img.ID
+			broadcaster, found := p.poolAdd("pull", poolKey)
+			broadcaster.Add(out)
 			if found {
-				broadcaster.Add(out)
-				broadcaster.Wait()
+				err = broadcaster.Wait()
```
Comment (cpuguy83) [inline graph/pull_v1.go]: Can this be `errors <- broadcaster.Wait()` and then do the error check on the receiver end?

File: graph/pull_v2.go [hunk 7]
```diff
@@ -243,6 +199,16 @@ func (p *v2Puller) pullV2Tag(out io.Writer, tag, taggedName string) (verified bo
 	var layerIDs []string
 	defer func() {
 		p.graph.Release(p.sessionID, layerIDs...)
+
+		for _, d := range downloads {
+			p.poolRemoveWithError("pull", d.poolKey, err)
+			if d.tmpFile != nil {
```
Comment (cpuguy83) [inline graph/pull_v2.go]: Helpful for reducing identation:
```
if d.tmpFile == nil {
    continue
} 
...
```

File: graph/pull_v1.go (hunk 6 above)
Comment (aaronlehmann) [inline graph/pull_v1.go]: Yes, I'll make this change. It was originally like this so that the `err == nil` case would display a "Download complete" message, but this message is redundant with the one that gets produced by the actual download.

File: graph/pull_v2.go (hunk 7 above)
Comment (aaronlehmann) [inline graph/pull_v2.go]: I agree with the suggestion in general, but in this particular case I'd rather leave it as-is, in case we add other cleanup tasks to this loop in the future (which would be skipped by the continue).

**Comment by aaronlehmann** [comment]
I suspect that the hanging test on the experimental build was caused by an issue with the test itself. It turns out that the `check` package uses `runtime.Goexit` when an assertion fails. This could mean that the `wg.Done()` call was never reached, and the waiting for the wait group hung forever. I've fixed these calls to use 'defer', which should cause them to execute even if the goroutine is exited in this way.

That's the good news. The bad news is that this line of reasoning would imply an assertion is failing. I'm not sure what would cause that to happen, and unfortunately I can't reproduce this locally. I'm going to try rebuilding this in CI a few times to see if an assertion failure ever happens (or if the the test hangs again).

**Comment by calavera** [comment]
@aaronlehmann our practice is to avoid using the check package inside goroutines. You can call the docker command directly and put the check after the goroutines:

https://github.com/docker/docker/blob/master/integration-cli/docker_utils.go#L636

**Comment by aaronlehmann** [comment]
I changed the tests to avoid using the `check` package in goroutines, and reran CI a few times. I'm not seeing any failures related to this PR, so I think this is ready to go if it looks good to reviewers.

**Comment by stevvooe** [comment]
@calavera Why would we avoid the use of the `check` package in goroutines? We should avoid unit testing frameworks that have odd limitations.

Typically, goroutines are not a problem with correctly formed check functions:

```
checkFoo(t *testing.T)
```

**Comment by jfrazelle** [comment]
can you squash the commits, then I think this looks like it has enough LGTM's from dist maintainers
